saexec/execution.go

```diff
@@ -0,0 +1,192 @@
+// Copyright (C) 2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package saexec
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"runtime"
+	"time"
+
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/vms/components/gas"
+	"github.com/ava-labs/libevm/common"
+	"github.com/ava-labs/libevm/core"
+	"github.com/ava-labs/libevm/core/state"
+	"github.com/ava-labs/libevm/core/state/snapshot"
+	"github.com/ava-labs/libevm/core/types"
+	"github.com/ava-labs/libevm/core/vm"
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/strevm/blocks"
+	"github.com/ava-labs/strevm/hook"
+)
+
+var errExecutorClosed = errors.New("saexec.Executor closed")
+
+// Enqueue pushes a new block to the FIFO queue. If [Executor.Close] is called
+// before [blocks.Block.Executed] returns true then there is no guarantee that
+// the block will be executed.
+func (e *Executor) Enqueue(ctx context.Context, block *blocks.Block) error {
+	for {
+		select {
+		case e.queue <- block:
+			return nil
+		case <-e.quit:
+			return errExecutorClosed
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			// If this happens then increase the channel's buffer size.
+			e.log.Warn("Execution queue buffer too small")
+			runtime.Gosched()
+		}
+	}
+}
+
+func (e *Executor) processQueue() {
+	defer close(e.done)
+
+	for {
+		select {
+		case <-e.quit:
+			return
+
+		case block := <-e.queue:
+			logger := e.log.With(
+				zap.Uint64("block_height", block.Height()),
+				zap.Uint64("block_time", block.BuildTime()),
+				zap.Stringer("block_hash", block.Hash()),
+				zap.Int("tx_count", len(block.Transactions())),
+			)
+
+			if err := e.execute(block, logger); err != nil {
+				logger.Error("Block execution failed", zap.Error(err))
+				return
+			}
+		}
+	}
+}
+
+type executionScratchSpace struct {
+	snaps   *snapshot.Tree
```

Comment:
Why is `snaps` part of the `executionScratchSpace`? `snaps` seems more similar to `stateCache` than it is to `statedb`.

Specifically, we never actually reassign `snaps`, and it isn't tied to a single `execute` call.

Reply:
From the early days of spiking on the PoC I was being pedantic about ownership and since only `stateCache` is documented as being thread safe I isolated the others. Now that we have much more reliable `blocks.Block.ParentBlock().PostExecutionStateRoot()`, the `StateDB` can be created as a local variable (as you suggest) so the scratch space is removed entirely.